### PR TITLE
test_volume_client: fix test_put_object_versioned()

### PR DESCRIPTION
**Pull request by rishabh-d-dave**
This PR...

* makes `test_put_object_versioned()` Python 3 compatible (with this all
  embedded Python programs in test_volume_client.py are Python 3 compatible)
* fixes the bug in `test_put_object_versioned()`
* renames `test_put_object_versioned()` since it's more than a simple test
* adds a basic positive test for `put_object_versioned()`

Besides,

* makes `sudo_write_file()` Python 3 compatible
* `_sudo_write_file()` is moved from test_volume_client.py to
  cephfs_test_case.py so that other  testsuites can use it too
* renames `_sudo_write_file()` to `sudo_write_file()` since other classes
  are encouraged to use it

- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

### CODE REVIEW

**Comment by rishabh-d-dave** [comment]
test_volume_client.TestVolumeClient.test_21501 failed - http://pulpito.ceph.com/rishabh-2019-06-24_04:33:11-fs-wip-rishabh-changes-to-test_put_object_versioned-distro-basic-smithi/

That's odd, because locally all tests in test_volume_client run successfully. Do I need to override [`py_version`](https://github.com/rishabh-d-dave/ceph/blob/changes-to-test_put_object_versioned/qa/tasks/cephfs/test_volume_client.py#L21) to `python3`? How can I do that?

**Comment by rishabh-d-dave** [comment]
I've pushed an extra commit that should force test_volume_client.py to use python3 for python programs embedded inside it. @batrick Is there a better to force teuthology to use python3? I sent an email asking this on CephFS mailing list but since I did not get any reply, I decided to go ahead with this for time being.

Link to the test logs - http://pulpito.front.sepia.ceph.com/rishabh-2019-06-28_14:59:50-fs-wip-rishabh-changes-to-test_put_object_versioned-distro-basic-smithi/

File: qa/tasks/cephfs/test_volume_client.py [hunk 1]
```diff
@@ -34,6 +34,7 @@ def _volume_client_python(self, client, script, vol_prefix=None, ns_prefix=None)
         return client.run_python("""
 from __future__ import print_function
 from ceph_volume_client import CephFSVolumeClient, VolumePath
+from rados import OSError as rados_OSError
```
Comment (sebastian-philipp) [inline qa/tasks/cephfs/test_volume_client.py]: does having a multiple similar exceptions `rados.OSError`, `cephfs.OSError` etc. actually provide any real benefit or is this just a weirdness?

**Comment by rishabh-d-dave** [comment]
Rishabh Dave [wrote](https://github.com/ceph/ceph/pull/28692#issuecomment-506768279):

The traceback from last test -
```
2019-07-01T04:42:05.486 INFO:tasks.cephfs_test_runner:======================================================================
2019-07-01T04:42:05.486 INFO:tasks.cephfs_test_runner:ERROR: test_21501 (tasks.cephfs.test_volume_client.TestVolumeClient)
2019-07-01T04:42:05.486 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2019-07-01T04:42:05.487 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2019-07-01T04:42:05.487 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-rishabh-changes-to-test_put_object_versioned/qa/tasks/cephfs/test_volume_client.py", line 1051, in test_21501
2019-07-01T04:42:05.487 INFO:tasks.cephfs_test_runner:    self._configure_vc_auth(vc_mount, "manila")
2019-07-01T04:42:05.487 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_wip-rishabh-changes-to-test_put_object_versioned/qa/tasks/cephfs/test_volume_client.py", line 61, in _configure_vc_auth
2019-07-01T04:42:05.487 INFO:tasks.cephfs_test_runner:    mount.client_remote.sudo_write_file(mount.get_keyring_path(), out)
2019-07-01T04:42:05.487 INFO:tasks.cephfs_test_runner:AttributeError: 'Remote' object has no attribute 'sudo_write_file'
```
I suppose, I need to make a modification equivalent to [this commit](https://github.com/ceph/ceph/pull/28692/commits/920b846bf4c9acb9620b96c56351d16b5e8f0b32) in teuthology repo too. @batrick Am I thinking in right direction?

File: qa/tasks/cephfs/test_volume_client.py [hunk 2]
```diff
@@ -989,21 +968,45 @@ def test_put_object_versioned(self):
...
+            vc.put_object("{pool_name}", "{obj_name}", data)
+            data = str.encode(data.decode() + 'm2')
+            try:
                 vc.put_object_versioned("{pool_name}", "{obj_name}", data, version)
-            """).format(pool_name=pool_name, obj_name=obj_name))
+            except {expected_exception}:
+                print('{expected_exception} raised')
+        """).format(pool_name=pool_name, obj_name=obj_name,
+                    expected_exception=expected_exception))
+        self.assertEqual(expected_exception + ' raised', output)
```
Comment (batrick) [inline qa/tasks/cephfs/test_volume_client.py]: I'm not sure why the other changes need to be present for this new test case. Can they be separate PRs?

File: qa/tasks/cephfs/test_volume_client.py [hunk 3]
```diff
@@ -14,7 +14,7 @@ class TestVolumeClient(CephFSTestCase):
     # One for looking at the global filesystem, one for being
     # the VolumeClient, two for mounting the created shares
     CLIENTS_REQUIRED = 4
-    default_py_version = 'python'
+    default_py_version = 'python3'
```
[the remark this answers is not in the document]
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_volume_client.py]: It's from a DNM commit that I wrote to make sure that this PR get tested against teuthology with Python 3 (See [this comment](https://github.com/ceph/ceph/pull/28692#issuecomment-506768279)). Your reply on how to test with Python3 came later.

File: qa/tasks/cephfs/test_volume_client.py (hunk 2 above)
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_volume_client.py]: I thought you agreed to merging these commits under one PR...

So, just the new test case should be moved? Because the code quoted above is just the old testcase (after improvement).

File: qa/tasks/cephfs/test_volume_client.py [hunk 4]
```diff
@@ -78,7 +58,7 @@ def _configure_vc_auth(self, mount, id_name):
             "mon", "allow *"
         )
         mount.client_id = id_name
-        self._sudo_write_file(mount.client_remote, mount.get_keyring_path(), out)
+        mount.client_remote.write_file(mount.get_keyring_path(), out)
```
[the remark this answers is not in the document]
Comment (rishabh-d-dave) [reply qa/tasks/cephfs/test_volume_client.py]: Using `sudo_write_file` from teuthology/misc.py now.

**Comment by rishabh-d-dave** [comment]
Link to the teuthology test - http://pulpito.front.sepia.ceph.com/rishabh-2019-07-10_04:13:34-fs-wip-rishabh-changes-to-test_put_object_versioned-distro-basic-smithi/

**Comment by rishabh-d-dave** [comment]
running tests using python2 - http://pulpito.front.sepia.ceph.com:80/rishabh-2019-07-12_17:59:04-fs-wip-rishabh-changes-to-test_put_object_versioned-distro-basic-smithi/

**Comment by rishabh-d-dave** [comment]
@batrick
Tests ran fine with both, python2[1] as well as python3[2] -
[1] http://pulpito.front.sepia.ceph.com/rishabh-2019-07-13_11:30:11-fs-wip-rishabh-changes-to-test_put_object_versioned-distro-basic-smithi/
[2] http://pulpito.front.sepia.ceph.com/rishabh-2019-07-13_11:29:15-fs-wip-rishabh-changes-to-test_put_object_versioned-distro-basic-smithi/
